src/protocol/codecs/payload.rs

```diff
@@ -0,0 +1,93 @@
+use std::io;
+
+use bytes::BytesMut;
+use tokio_util::codec::Decoder;
+use tracing::Span;
+
+use crate::protocol::{
+    codecs::{
+        tagmsg::Tag,
+        topic::{Topic, TopicCodec},
+    },
+    invalid_data,
+};
+
+#[derive(Debug)]
+pub enum Payload {
+    MsgOfInterest(MsgOfInterest),
+    NotImplemented,
+}
+
+/// [MsgOfInterest] contains a tag list in which the node is interested.
+#[derive(Debug)]
+pub struct MsgOfInterest {
+    /// Message tags for which the node is interested (subscribed).
+    // TODO(Rqnsom): convert to a hashset.
+    pub tags: String,
+}
+
+impl MsgOfInterest {
+    /// Constructs a [MsgOfInterest] payload from the [topics] list.
+    pub fn construct(mut topics: Vec<Topic>) -> Result<Self, io::Error> {
+        if topics.len() != 1 {
+            return Err(invalid_data!("expected a single topic"));
+        }
+
+        let tag_topic = topics.pop().unwrap();
+        if tag_topic.key != "tags" {
+            return Err(invalid_data!("expected 'tags' topic"));
+        }
+
+        Ok(Self {
+            tags: tag_topic.value,
+        })
+    }
+}
+
+/// [PayloadCodec] decodes the Algod message payload using a provided tag.
+#[derive(Clone)]
+pub struct PayloadCodec {
+    /// The associated node's span.
+    span: Span,
+
+    /// Represents a message payload type identifier.
+    pub tag: Option<Tag>,
```

Comment:
I'd add a (non-doc) comment here explaining why the option is used

Reply:
Done